### PR TITLE
[TTS] Add Mixed Representation Training

### PR DESCRIPTION
**Pull request by blisc**
- Adds phoneme probability to our current g2p module.
- Updates FastPitch yaml to use this functionality.

### CODE REVIEW

**Comment by Oktai15** [comment]
See this bug: https://github.com/NVIDIA/NeMo/pull/3593

**Comment by Oktai15** [comment]
@blisc

File: nemo/collections/tts/torch/tts_tokenizers.py [hunk 1]
```diff
@@ -282,6 +282,9 @@ def __init__(
              Note that lower() function shouldn't applied here, because text can contains phonemes (it will be handled by g2p).
```
Comment (Oktai15) [inline nemo/collections/tts/torch/tts_tokenizers.py:282]: Current implementation of `EnglishPhonemesTokenizer` with `phoneme_probability > 0` in G2P also mixes representation even in eval mode. We don't need it, right? Let's add flag to `EnglishPhonemesTokenizer.encode` like `eval_mode`. Then in `encode` method you need to replace this line `g2p_text = self.g2p(text)` with something like that `g2p_text = self.g2p(text, mix_representation=False if eval_mode else None)`. And then you need to add `mix_representation` in g2p and handle it. Also we need to update https://github.com/blisc/NeMo/blob/tts_finetuning_updates/nemo/collections/tts/models/fastpitch.py#L197 somehow and, please, also update Mixer-TTS function (should be the same changes).

**Review by Oktai15 — CHANGES_REQUESTED** [review]
Great PR, thanks for this feature. Need to fix comments.

File: examples/tts/conf/fastpitch_align_v1.05.yaml [hunk 2]
```diff
@@ -10,8 +10,8 @@ sup_data_path: ???
 sup_data_types: [ "align_prior_matrix", "pitch" ]
```
[the remark this answers is not in the document]
Comment (blisc) [reply examples/tts/conf/fastpitch_align_v1.05.yaml:10]: I think this will be the default fastpitch config moving forward as it is what will be used in riva.

File: examples/tts/conf/fastpitch_align_v1.05.yaml [hunk 3]
```diff
@@ -101,12 +102,13 @@ model:
...
       pitch_mean: ${model.pitch_mean}
       pitch_std: ${model.pitch_std}
+      use_beta_binomial_interpolator: true
 
     dataloader_params:
       drop_last: false
       shuffle: true
       batch_size: 32
-      num_workers: 12
+      num_workers: 0
```
[the remark this answers is not in the document]
Comment (blisc) [reply examples/tts/conf/fastpitch_align_v1.05.yaml:111]: It's not an OOM error. I am still unsure why there is an error. I'm not sure if num_workers=0 is stable even.

File: nemo/collections/tts/torch/data.py [hunk 4]
```diff
@@ -179,15 +183,15 @@ def __init__(
...
-
                         if self.text_normalizer is not None:
                             text = self.text_normalizer_call(text, **self.text_normalizer_call_kwargs)
-
                         file_info["normalized_text"] = text
-                        file_info["text_tokens"] = self.text_tokenizer(text)
                     else:
                         file_info["normalized_text"] = item["normalized_text"]
-                        file_info["text_tokens"] = self.text_tokenizer(item["normalized_text"])
+                        file_info["raw_text"] = item["text"]
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/data.py:191]: Removed

File: nemo/collections/tts/torch/data.py [hunk 5]
```diff
@@ -331,6 +336,8 @@ def add_align_prior_matrix(self, **kwargs):
         self.align_prior_matrix_folder.mkdir(exist_ok=True, parents=True)
 
         self.use_beta_binomial_interpolator = kwargs.pop('use_beta_binomial_interpolator', False)
+        if not self.cache_text:
+            self.use_beta_binomial_interpolator = True
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/data.py:339]: Added something similar.

File: nemo/collections/tts/torch/data.py [hunk 6]
```diff
@@ -417,16 +428,16 @@ def __getitem__(self, index):
         # Load alignment prior matrix if needed
         align_prior_matrix = None
         if AlignPriorMatrix in self.sup_data_types_set:
+            mel_len = self.get_log_mel(audio).shape[2]
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/data.py:430]: reverted

File: nemo/collections/tts/torch/g2ps.py [hunk 7]
```diff
@@ -67,6 +70,7 @@ def __init__(
             ignore_ambiguous_words: Whether to not handle word via phoneme_dict with ambiguous phoneme sequences. Defaults to True.
             heteronyms (str, Path, List): Path to file with heteronyms (every line is new word) or list of words.
             encoding: Encoding type.
+            phoneme_probability (Optional[float]): The probability (0<var<1) that each word is phonemized. Defaults to None which is the same as 100%.
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/g2ps.py:73]: updated

File: nemo/collections/tts/torch/g2ps.py [hunk 8]
```diff
@@ -163,6 +169,9 @@ def parse_one_word(self, word: str):
         `status` will be `False` if word wasn't handled, `True` otherwise.
         """
 
+        if self.phoneme_probability is not None and self._rng.random() > self.phoneme_probability:
+            return word, False
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/g2ps.py:173]: Updated

File: examples/tts/conf/fastpitch_align_v1.05.yaml [hunk 9]
```diff
@@ -10,8 +10,8 @@ sup_data_path: ???
 sup_data_types: [ "align_prior_matrix", "pitch" ]
 
 # Default values from librosa.pyin
-pitch_fmin: 65.40639132514966
-pitch_fmax: 2093.004522404789
+pitch_fmin: 80
```
[the remark this answers is not in the document]
Comment (blisc) [reply examples/tts/conf/fastpitch_align_v1.05.yaml:13]: reverted

File: nemo/collections/tts/torch/tts_tokenizers.py [hunk 10]
```diff
@@ -295,7 +298,7 @@ def __init__(
             vowels = [f'{p}{s}' for p, s in itertools.product(vowels, (0, 1, 2))]
         tokens.extend(vowels)
 
-        if chars:
+        if chars or self.phoneme_probability is not None:
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/tts_tokenizers.py:301]: Added

File: nemo/collections/tts/torch/tts_tokenizers.py (hunk 1 above)
Comment (blisc) [reply nemo/collections/tts/torch/tts_tokenizers.py:282]: Do you want it to be run in eval()?
I can enable for train, and disable for infer.

Comment (Oktai15) [reply nemo/collections/tts/torch/tts_tokenizers.py:282]: I want to have ability to turn off stochastic mixed representation if I need it. I can need it during evaluation_step(), also I can need it in `def self.parse()`…

File: nemo/collections/tts/torch/data.py [hunk 11]
```diff
@@ -336,6 +336,11 @@ def add_align_prior_matrix(self, **kwargs):
 
         self.use_beta_binomial_interpolator = kwargs.pop('use_beta_binomial_interpolator', False)
         if not self.cache_text:
+            if 'use_beta_binomial_interpolator' in kwargs and not self.use_beta_binomial_interpolator:
+                logging(
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/torch/data.py]: fixed

File: Jenkinsfile [hunk 12]
```diff
@@ -2177,11 +2177,10 @@ pipeline {
...
         }
-        // TODO(Oktai15): update it in 1.8.0 version
         stage('FastPitch') {
           steps {
             sh 'python examples/tts/fastpitch.py \
-            --config-name fastpitch_align \
+            --config-name fastpitch_align_v1.05 \
             train_dataset=/home/TestData/an4_dataset/an4_train.json \
             validation_datasets=/home/TestData/an4_dataset/an4_val.json \
             prior_folder=/home/TestData/an4_dataset/beta_priors \
```
Comment (Oktai15) [inline Jenkinsfile:2186]: remove it and add `sup_data_path `

File: nemo/collections/tts/models/fastpitch.py [hunk 13]
```diff
@@ -195,13 +195,21 @@ def parser(self):
         return self._parser
 
     def parse(self, str_input: str, normalize=True) -> torch.tensor:
+        if self.training:
+            logging.warning("parse() is meant to be called in eval mode.")
         if str_input[-1] not in [".", "!", "?"]:
```
Comment (Oktai15) [inline nemo/collections/tts/models/fastpitch.py:200]: Let's remove it?

File: nemo/collections/tts/models/fastpitch.py [hunk 14]
```diff
@@ -195,13 +195,21 @@ def parser(self):
...
             str_input = str_input + "."
 
         if normalize and self.text_normalizer_call is not None:
             str_input = self.text_normalizer_call(str_input, **self.text_normalizer_call_kwargs)
 
-        tokens = self.parser(str_input)
+        if self.learn_alignment:
+            # Disable mixed g2p representation
+            with self.vocab.set_phone_prob():
```
Comment (Oktai15) [inline nemo/collections/tts/models/fastpitch.py:208]: Good idea. Could we set explicitly `prob=1.0` or rename it to `self.vocab.disable_mixed_representation()`, because I don't have idea in which case we can run this context manager except this one.

File: nemo/collections/tts/models/fastpitch.py [hunk 15]
```diff
@@ -195,13 +195,21 @@ def parser(self):
         return self._parser
 
     def parse(self, str_input: str, normalize=True) -> torch.tensor:
+        if self.training:
+            logging.warning("parse() is meant to be called in eval mode.")
```
Comment (Oktai15) [inline nemo/collections/tts/models/fastpitch.py:199]: Great idea. Let's do the same in `generate_spectrogram`:

Remove it:
```
# FIXME: return masks as well?
 self.eval()
 ```
 
 And add it:
 
 ```
 if self.training:
            logging.warning("generate_spectrogram() is meant to be called in eval mode.")
 ````

File: nemo/collections/tts/models/mixer_tts.py [hunk 16]
```diff
@@ -644,9 +644,13 @@ def generate_spectrogram(
         return pred_spect
 
     def parse(self, text: str, normalize=True) -> torch.Tensor:
```
Comment (Oktai15) [inline nemo/collections/tts/models/mixer_tts.py:646]: Same comments as for fastpitch

File: nemo/collections/tts/torch/g2ps.py [hunk 17]
```diff
@@ -67,6 +70,7 @@ def __init__(
             ignore_ambiguous_words: Whether to not handle word via phoneme_dict with ambiguous phoneme sequences. Defaults to True.
             heteronyms (str, Path, List): Path to file with heteronyms (every line is new word) or list of words.
             encoding: Encoding type.
+            phoneme_probability (Optional[float]): The probability (0.<var<1.) that each word is phonemized. Defaults to None which is the same as 1.
```
Comment (Oktai15) [inline nemo/collections/tts/torch/g2ps.py:73]: The probability (0.<var<1.) that each word is phonemized. -> The probability (0.<var<1.) that each word is phonemized if it is possible.

Because this tokenizer returns chars too.

File: nemo/collections/tts/torch/tts_tokenizers.py (hunk 1 above)
Comment (Oktai15) [reply nemo/collections/tts/torch/tts_tokenizers.py:283]: It is not good that our metrics in validation step will be depend on `phoneme_probability`, because they will be not representative for user (in inference user will use `parse` and it produces deterministic output). 

How to fix it? 

I see two variants:

1. In `__setup_dataloader_from_config` for `name="val"` we can do something like that:

 ```
 val_vocab = copy.deepcopy(self.vocab) # or any other alternative to copy.deepcopy
 val_vocab. phoneme_probability = 1.0
 val_vocab.g2p. phoneme_probability = 1.0 # need to check if g2p is
 dataset = instantiate(...text_tokenizer=val_vocab)
 ```
 
 We need to do the same for Mixer-TTS.
 
 2. We can add field `self.dataset_mode` to TTSDataset  (it can be "val_dataset" or "train_dataset"). In this case, we can add your context manager if `self.dataset_mode == val_dataset`

**Review by Oktai15 — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/models/fastpitch.py (hunk 14 above)
Comment (blisc) [reply nemo/collections/tts/models/fastpitch.py]: Made prob required.

File: nemo/collections/tts/torch/g2ps.py (hunk 17 above)
Comment (blisc) [reply nemo/collections/tts/torch/g2ps.py:73]: Updated comment

File: nemo/collections/tts/torch/tts_tokenizers.py (hunk 1 above)
Comment (blisc) [reply nemo/collections/tts/torch/tts_tokenizers.py:283]: Added a fix to fastpitch. Please update mixerTTS as you want

File: nemo/collections/tts/models/fastpitch.py (hunk 15 above)
Comment (blisc) [reply nemo/collections/tts/models/fastpitch.py:199]: Done

File: nemo/collections/tts/models/fastpitch.py [hunk 18]
```diff
@@ -254,8 +254,8 @@ def forward(
 
     @typecheck(output_types={"spect": NeuralType(('B', 'D', 'T_spec'), MelSpectrogramType())})
     def generate_spectrogram(self, tokens: 'torch.tensor', speaker: int = 0, pace: float = 1.0) -> torch.tensor:
-        # FIXME: return masks as well?
-        self.eval()
+        if self.training:
+            logging.warning("parse() is meant to be called in eval mode.")
```
Comment (Oktai15) [inline nemo/collections/tts/models/fastpitch.py:258]: parse —> generate_spectogram

**Review by Oktai15 — CHANGES_REQUESTED** [review]

**Review by Oktai15 — APPROVED** [review]
Great PR, thanks.

We need to think more about accuracy of beta binomial interpolator and also think about potential bug in TTSDataset
